PP-516: Additional tests for direct write and remove files

visheshh opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-516](https://pbspro.atlassian.net/browse/PP-516)**

#### Solution description
Additional tests for direct write and remove files

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

minghui-liu commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -64,6 +64,126 @@ def test_direct_write_when_job_succeeds(self):
+        directory when direct_files option is used with o option.
+        """
+        j = Job(TEST_USER)
+        j.set_attributes({ATTR_k: 'do'})
```
Please use `j = Job(TEST_USER, attrs={ATTR_k: 'do'})` to combine the above two lines into one.

minghui-liu commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -64,6 +64,126 @@ def test_direct_write_when_job_succeeds(self):
+    def test_direct_write_output_file(self):
+        """
+        submit a sleep job and make sure that the output file
+        are getting directly written directly to the mapped
```
There are two "directly" in this sentence. Please remove one. Please also fix the comments of the other functions.

minghui-liu commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -64,6 +64,126 @@ def test_direct_write_when_job_succeeds(self):
+        j.set_attributes({ATTR_e: err_file})
+        j.set_attributes({ATTR_o: out_file})
+        j.set_attributes({ATTR_k: 'doe'})
+        j.set_sleep_time(10)
```
Please combine the attributes into a dictionary and pass it to the `Job` call.

minghui-liu commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -64,6 +64,126 @@ def test_direct_write_when_job_succeeds(self):
+        j.set_attributes({ATTR_e: tmp_dir})
+        j.set_attributes({ATTR_o: tmp_dir})
+        j.set_attributes({ATTR_k: 'doe'})
+        j.set_sleep_time(10)
```
Same as above. You can do
```
attr = {
  ATTR_e: tmp_dir,
  ATTR_o: tmp_dir,
  ATTR_k: 'doe'
}
j = Job(TEST_USER, attrs=attr)
```

minghui-liu commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -106,3 +226,49 @@ def test_qalter_direct_write(self):
+        """
+        self.mom.add_config({'$logevent': '0xffffffff'})
+        j = Job(TEST_USER)
+        j.set_attributes({ATTR_k: 'doe'})
```
Same as above. You do not need a separate `set_attributes` call. Please also fix other places.

minghui-liu commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -64,6 +64,126 @@ def test_direct_write_when_job_succeeds(self):
+        jid = self.server.submit(j, submit_dir=sub_dir)
+        self.server.log_match(jid + ";Job Run at", max_attempts=10)
+        file_count = len([name for name in os.listdir(
+            mapping_dir) if os.path.isfile(os.path.join(mapping_dir, name))])
```
Python has a `filter` function that filters a list based on a condition. 
```
is_file = lambda name : os.path.isfile(os.path.join(mapping_dir, name))
file_count = len(filter(is_file, os.listdir(mapping_dir))
```
I think it is a bit more readable.

visheshh replied:
autopep8 converted this into 
def is_file(name): return os.path.isfile(
            os.path.join(mapping_dir, name))
and complained about having 2 functions , May be i'll keep it the way it is.

nithinj commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -64,6 +64,126 @@ def test_direct_write_when_job_succeeds(self):
+        """
+        j = Job(TEST_USER)
+        j.set_attributes({ATTR_k: 'do'})
+        j.set_sleep_time(10)
```
You can reduce the sleep time to 1 sec to make the test case faster.

nithinj commented on test/tests/functional/pbs_qsub_remove_files.py:
```diff
@@ -54,6 +54,123 @@ def test_remove_file_when_job_succeeds(self):
+        self.server.log_match(jid + ";dequeuing from", max_attempts=10)
+        file_count = len([name for name in os.listdir(
+            sub_dir) if os.path.isfile(os.path.join(sub_dir, name))])
+        self.assertEqual(1, file_count)
```
This test makes sure that only one file exists at the end of the job. But we need to make sure that it is the error file.

nithinj commented on test/tests/functional/pbs_qsub_remove_files.py:
```diff
@@ -54,6 +54,123 @@ def test_remove_file_when_job_succeeds(self):
+        self.server.log_match(jid + ";dequeuing from", max_attempts=10)
+        file_count = len([name for name in os.listdir(
+            mapping_dir) if os.path.isfile(os.path.join(mapping_dir, name))])
+        self.assertEqual(1, file_count)
```
This test makes sure that only one file exists at the end of the job. But we need to make sure that it is the output file.

visheshh commented:
Hi @nithinj and @minghui-liu , I have addressed review comments please review .
I have not used lambda function because autopep8 complains of a function inside a function.

visheshh commented:
Hi @minghui-liu , Please review the code changes .

visheshh commented:
@PBSPro/pbspro-maintainers , Please review the PR .

arungrover commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -47,12 +47,136 @@ class TestQsub_direct_write(TestFunctional):
+        when direct_files option is used.
+        """
+        j = Job(TEST_USER, attrs={ATTR_k: 'doe'})
+        j.set_sleep_time(10)
```
you can reduce the sleep time to a second rather than 10 seconds because copying error/out files is all you are looking to test

visheshh replied:
Hi @arungrover , I am checking file counts for checking for directly written files .
For Example: If job completes within 1 second then both directly written and staged files will be in mapping directory once the job gets completed .

arungrover replied:
In this case please write a comment why longer sleep times are necessary.

bhroam replied:
Instead of a comment, I'd write it in a log message.  It'd help the person running the test to understand why the test is sleeping.

arungrover commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -106,3 +230,48 @@ def test_qalter_direct_write(self):
+            self.server.alterjob(jid, attribs)
+        except PbsAlterError as e:
+            self.assertTrue(
+                'qalter: Cannot modify attribute while job running  Keep_Files'
```
are you sure that when run is API mode it will spit out an error prepended with "qalter: " in  it?
Just remove it if it is not important. Are you also sure there will be 2 spaces between the error message and attribute name ("Keep_Files")?

visheshh replied:
Yes there are 2 spaces

arungrover commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -106,3 +230,48 @@ def test_qalter_direct_write(self):
+        self.server.expect(JOB, {'Keep_Files': 'doe'}, id=jid)
+        self.server.log_match(jid + ";Job Run at", max_attempts=10)
+        self.server.rerunjob(jid)
+        self.mom.log_match(
```
Don't you want to specify the starttime while doing this log match?

arungrover commented on test/tests/functional/pbs_qsub_remove_files.py:
```diff
@@ -49,11 +49,130 @@ def test_remove_file_when_job_succeeds(self):
+        sub_dir = DshUtils().mkdtemp(uid=TEST_USER.uid)
+        jid = self.server.submit(j, submit_dir=sub_dir)
+        self.server.expect(JOB, {'Remove_Files': 'oe'}, id=jid)
+        self.server.log_match(jid + ";dequeuing from", max_attempts=10)
```
can't we make sure that the job has finished instead of relying on this message?

visheshh replied:
Hi @arungrover , Do you want to me make history enabled and check for F state ?

arungrover replied:
That can work too, or just check for non-existence of job and then proceed

arungrover commented on test/tests/functional/pbs_qsub_remove_files.py:
```diff
@@ -99,3 +217,20 @@ def test_qalter_remove_files(self):
+            self.server.alterjob(jid, attribs)
+        except PbsAlterError as e:
+            self.assertTrue(
+                'qalter: Cannot modify attribute while job'
```
You might want to remove "qalter: " from the message you are trying to search because it may not be there if PTL is running in API mode.

visheshh commented on test/tests/functional/pbs_qsub_direct_write.py:
```diff
@@ -47,12 +47,136 @@ class TestQsub_direct_write(TestFunctional):
+        file_count = len([name for name in os.listdir(
+            mapping_dir) if os.path.isfile(os.path.join(mapping_dir, name))])
+        self.assertEqual(2, file_count)
+        self.server.expect(JOB, {'Keep_Files': 'doe'}, id=jid)
```
As doe is a new value , I am just testing it is set . May be i can just do it for one of the tests

visheshh commented:
Hi @arungrover , I have addressed all your review comments , Please review

visheshh commented:
squashed and rebased , Please take this PR forward.
